File: Year 2/Semester 4/MPP/ProtobuffProjectCSharp/networking/ClientRpcReflectionWorker.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization.Formatters.Binary;
using System.Runtime.Serialization;
using System.Threading;
using ProtobuffProject.services;
using ProtobuffProject.domain;
using System.Reflection;
using log4net;
using System.IO;

namespace ProtobuffProject.networking
{
    public class ClientRpcReflectionWorker : IObserver //, Runnable
    {

        private static readonly ILog Logger = LogManager.GetLogger("ClientRpcReflectionWorker");

        private readonly IServices service;
        private readonly TcpClient connection;
        private readonly NetworkStream stream;
        private readonly IFormatter formatter;
        private volatile bool connected;
        public ClientRpcReflectionWorker(IServices service, TcpClient connection)
        {
            this.service = service;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void Run()
        {
            while (connected)
            {
                try
                {
                    var request = formatter.Deserialize(stream);
                    object response = HandleRequest((Request)request);
                    if (response != null)
                    {
                        SendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private Response HandleRequest(Request request)
        {       
            Response response = null;
            var handlerName = "Handle" + request.Type;
            Console.WriteLine("HandlerName " + handlerName);
            try
            {
                var method = this.GetType().GetMethod(handlerName, new[] { typeof(Request) });
                response = (Response)method.Invoke(this, new[] { request });
                Console.WriteLine("Method " + handlerName + " invoked");
            }
            catch (TargetInvocationException ex)    
            {
                Console.WriteLine(ex.InnerException);
            }
            catch (AmbiguousMatchException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (TargetParameterCountException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (TargetException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            return response;
        }

        public Response HandleLOGIN(Request request)
        {
            Logger.InfoFormat("method entered: handleLOGIN with parameters {0}", request);
            var employee = (Employee<int>)request.Data;
            try
            {
                var foundEmployee = service.AuthenticateEmployee(employee, this);
                Logger.Info("Employee logged in");
                return new Response.Builder().Type(ResponseType.OK).Data(foundEmployee).Build();
            }
            catch (MyException e)
            {
                connected = false;
                Logger.Error("Error in worker (solving method handleLOGIN): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleBUY_TICKET(Request request)
        {
            Logger.InfoFormat("method entered: handleBUY_TICKET with parameters {0}", request);
            var ticket = (Ticket<int>)request.Data;
            try
            {
                service.AddTickets(ticket);
                Logger.Info("Ticket bought");
                return new Response.Builder().Type(ResponseType.OK).Data(ticket).Build();
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleBUY_TICKET): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleLOGOUT(Request request)
        {
            Logger.InfoFormat("method entered: handleLOGOUT with parameters {0}", request);
            var employee = (Employee<int>)request.Data;
            try
            {
                service.Logout(employee);
                connected = false;
                Logger.Info("User logged out");
                return new Response.Builder().Type(ResponseType.OK).Build(); ;
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleLOGOUT): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleGET_ALL_SHOWS(Request request)
        {
            Logger.InfoFormat("method entered: handleGET_ALL_SHOWS with parameters {0}", request);
            Console.WriteLine("HandleGetAllShows");
            try
            {
                var showsDTO = (List<ShowDTO<int>>)service.GetAllShowsDTO();
                Logger.InfoFormat("Shows found {0}", showsDTO);
                Console.WriteLine(@"Shows found {0}", showsDTO);
                return new Response.Builder().Type(ResponseType.GET_ALL_SHOWS).Data(showsDTO).Build();
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleGET_ALL_SHOWS): {0}", e);
                Console.WriteLine(@"Error in worker (solving method handleGET_ALL_SHOWS): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleGET_ALL_ARTISTS(Request request)
        {
            Logger.InfoFormat("method entered: handleGET_ALL_ARTISTS with parameters {0}", request);
            try
            {
                var artists = (List<Artist<int>>)service.GetAllArtists();
                Logger.InfoFormat("Artists found {0}", artists);
                return new Response.Builder().Type(ResponseType.GET_ALL_ARTISTS).Data(artists).Build();
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleGET_ALL_ARTISTS): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleFIND_ARTIST(Request request)
        {
            Logger.InfoFormat("method entered: handleFIND_ARTIST with parameters {0}", request);
            Console.WriteLine("method entered: handleFIND_ARTIST with parameters " + request);
            var idArtist = (int)request.Data;
            try
            {
                var artist = service.FindArtist(idArtist);
                Logger.InfoFormat("Artist found {0}", artist);
                Console.WriteLine("Artist found " + artist);
                return new Response.Builder().Type(ResponseType.FIND_ARTIST).Data(artist).Build();
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleFIND_ARTIST): {0}", e);
                Console.WriteLine("Error in worker (solving method handleFIND_ARTIST): " + e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();          
            }    
        }

        private void SendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            lock (stream)
            {
                formatter.Serialize(stream, response);
                stream.Flush();
            }
        }

        public void TicketAdded(Ticket<int> ticket)
        {
            var resp = new Response.Builder().Type(ResponseType.TICKET_SOLD).Data(ticket).Build();
            Logger.Info($"Ticket sold {ticket}");
            Console.WriteLine($@"Ticket sold {ticket}");
            try
            {
                SendResponse(resp);
                Logger.Info("Response sent");
                Console.WriteLine("Response sent");
            }
            catch (IOException e)
            {
                Console.WriteLine(e);
                Logger.Error($"Error in worker (sending response): {e}");
                Console.WriteLine(@"Error in worker (sending response): {0}", e);
            }
        }
    }

}

File: Year 2/Semester 4/Programming and Projection Environments/Week 9/ProtobuffProjectJava/Networking/src/main/java/gRPCProject/network/rpcprotocol/ClientRpcReflectionWorker.java
package gRPCProject.network.rpcprotocol;

import gRPCProject.domain.*;
import gRPCProject.service.IObserver;
import gRPCProject.service.IService;
import gRPCProject.service.MyException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.util.List;


public class ClientRpcReflectionWorker implements Runnable, IObserver {
    private final IService service;
    private final Socket connection;

    private static final Logger logger= LogManager.getLogger();

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;
    public ClientRpcReflectionWorker(IService service, Socket connection) {
        logger.info("Creating worker");
        this.service = service;
        this.connection = connection;
        try{
            output=new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input=new ObjectInputStream(connection.getInputStream());
            connected=true;
            logger.info("Worker created");
        } catch (IOException e) {
            logger.error(e);
        }
    }

    public void run() {
        while(connected){
            try {
                Object request=input.readObject();
                Response response=handleRequest((Request)request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException | ClassNotFoundException e) {
                logger.error("Error in worker (reading): "+e);
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                logger.error("Error in worker (sleeping): "+e);
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error in worker (closing connection): "+e);
        }
    }

    private static final Response okResponse=new Response.Builder().type(ResponseType.OK).build();
    private Response handleRequest(Request request){
        Response response=null;
        String handlerName="handle"+(request).type();
        logger.traceEntry("method entered: "+handlerName+" with parameters "+request);
        try {
            Method method=this.getClass().getDeclaredMethod(handlerName, Request.class);
            response=(Response)method.invoke(this,request);
            logger.info("Method invoked: "+handlerName+" with response "+response);
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
            e.printStackTrace();
            logger.error("Error in worker (invoking method handleRequest): "+e);
        }
        return response;
    }

    private Response handleLOGIN(Request request){
        logger.traceEntry("method entered: handleLOGIN with parameters "+request);
        Employee employee = (Employee) request.data();
        try {
            Employee foundEmployee = service.authenticateEmployee(employee, this);
            logger.info("Employee logged in");
            return new Response.Builder().type(ResponseType.OK).data(foundEmployee).build();
        } catch (MyException e) {
            connected=false;
            logger.error("Error in worker (solving method handleLOGIN): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleBUY_TICKET(Request request){
        logger.traceEntry("method entered: handleBUY_TICKET with parameters "+request);
        Ticket ticket=(Ticket) request.data();
        try {
            service.addTickets(ticket);
            logger.info("Ticket bought");
            return new Response.Builder().type(ResponseType.OK).data(ticket).build();
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleBUY_TICKET): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }
    private Response handleLOGOUT(Request request){
        logger.traceEntry("method entered: handleLOGOUT with parameters "+request);
        Employee employee=(Employee) request.data();
        try {
            service.logout(employee, this);
            connected=false;
            logger.info("User logged out");
            return okResponse;
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleLOGOUT): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_ALL_SHOWS(Request request){
        logger.traceEntry("method entered: handleGET_ALL_SHOWS with parameters "+request);
        try {
            List<ShowDTO> showsDTO= (List<ShowDTO>) service.getAllShowsDTO();
            logger.info("Shows found "+showsDTO);
            return new Response.Builder().type(ResponseType.GET_ALL_SHOWS).data(showsDTO).build();
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleGET_ALL_SHOWS): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_ALL_ARTISTS(Request request){
        logger.traceEntry("method entered: handleGET_ALL_ARTISTS with parameters "+request);
        try {
            List<Artist> artists= (List<Artist>) service.getAllArtists();
            logger.info("Artists found "+artists);
            return new Response.Builder().type(ResponseType.GET_ALL_ARTISTS).data(artists).build();
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleGET_ALL_ARTISTS): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleFIND_ARTIST(Request request){
        logger.traceEntry("method entered: handleFIND_ARTIST with parameters "+request);
        Integer idArtist=(Integer) request.data();
        try {
            Artist artist=service.findArtist(idArtist);
            logger.info("Artist found "+artist);
            return new Response.Builder().type(ResponseType.FIND_ARTIST).data(artist).build();
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleFIND_ARTIST): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private void sendResponse(Response response) throws IOException{
        logger.traceEntry("method entered: sendResponse with parameters "+response);
        output.writeObject(response);
        output.flush();
        logger.info("Response sent");
    }

    @Override
    public void ticketAdded(Ticket ticket) throws MyException {
        Response resp=new Response.Builder().type(ResponseType.TICKET_SOLD).data(ticket).build();
        logger.info("Ticket sold "+ticket);
        try {
            sendResponse(resp);
            logger.info("Response sent");
        } catch (IOException e) {
            e.printStackTrace();
            logger.error("Error in worker (sending response): "+e);
        }
    }
}
